### PR TITLE
EMKTG-3254/fix/filter estimate carriers

### PR DESCRIPTION
**Pull request by ClaraLpresta**

### CODE REVIEW

File: _dev/apps/ui/src/store/modules/product-feed/actions.ts [hunk 1]
```diff
@@ -50,6 +51,18 @@ export const createProductFeedApiPayload = (settings:any) => ({
   requestSynchronizationNow: settings.requestSynchronizationNow,
 });
 
+const filterEstimateCarriers = (data) => {
```
Comment (kktos) [inline _dev/apps/ui/src/store/modules/product-feed/actions.ts:54]: you should add data type here

**Review by kktos — APPROVED** [review]

File: .gitignore [hunk 2]
```diff
@@ -34,4 +34,4 @@ _dev/apps/ui/.jest-test-results.json
 _dev/apps/ui/storybook-static
 .vscode/launch.json
 .pnpm-store
-views/js/fetchVerificationTag.js*
+views/js/*
```
Comment (Quetzacoalt91) [inline .gitignore:37]: This line is too aggressive in my opinion because some files are currently tracked in this folder.
More files could be tracked in the future as well, and with this line there would never appear in the files list ready to be staged.

I would rather manually set the list of files to ignore.

File: _dev/apps/ui/src/store/modules/product-feed/actions.ts [hunk 3]
```diff
@@ -50,6 +52,18 @@ export const createProductFeedApiPayload = (settings:any) => ({
...
 
+const filterEstimateCarriers = (data: CustomCarrier[]) => {
+  if (data[0].offer === OfferType.FLAT_SHIPPING_RATE || data[0].offer === OfferType.FREE_SHIPPING) {
+    delete data[0].freeShippingOverAmount;
+    if (data[0].offer === OfferType.FREE_SHIPPING) {
+      delete data[0].flatShippingRate;
+    }
+  } else if (data[0].offer === OfferType.FREE_SHIPPING_OVER_AMOUNT) {
+    delete data[0].flatShippingRate;
+  }
```
Comment (Quetzacoalt91) [inline _dev/apps/ui/src/store/modules/product-feed/actions.ts:63]: This change will fix the first element of the array. It is possible to estimate several carriers in the product feed, for instance by having a carrier by country.

You should apply your fix to each line.

**Review by Quetzacoalt91 — CHANGES_REQUESTED** [review]

File: _dev/apps/ui/.editorconfig [hunk 4]
```diff
@@ -3,3 +3,4 @@ indent_style = space
 indent_size = 2
 trim_trailing_whitespace = true
 insert_final_newline = true
+end_of_line = lf
```
[the remark this answers is not in the document]
Comment (kktos) [reply _dev/apps/ui/.editorconfig:6]: Clara & I got the linter complaining that all the files are CRLF ended while it expects LF.
Trying to fix this.
this is the reason why ;)

[the remark this answers is not in the document]
Comment (Quetzacoalt91) [reply _dev/apps/ui/.editorconfig:6]: All right! I expected many files to be changed to definitely fix this, this should be in a dedicated PR.

**Review by kktos — DISMISSED** [review]

**Review by kktos — APPROVED** [review]

File: _dev/apps/ui/src/store/modules/product-feed/actions.ts [hunk 5]
```diff
@@ -1,54 +1,81 @@
...
-  ...(
-    (settings.shippingSetup === ShippingSetupOption.IMPORT) ? {
+export const createProductFeedApiPayload = (settings: ProductFeedAPIPayload): ProductFeedAPIPayload => {
+  // const productFeedApiPayload: Partial<ProductFeedAPIPayload>= {
+  const productFeedApiPayload: ProductFeedAPIPayload = {
+    autoImportTaxSettings: settings.autoImportTaxSettings,
+    shippingSetup: settings.shippingSetup,
+    targetCountries: settings.targetCountries,
+
+    // biome-ignore lint/suspicious/noExplicitAny: TYPE:MUST_FIX
```
Comment (Quetzacoalt91) [inline _dev/apps/ui/src/store/modules/product-feed/actions.ts:36]: Not related to the linter we use in the project

File: _dev/apps/ui/src/store/modules/product-feed/actions.ts [hunk 6]
```diff
@@ -1,54 +1,81 @@
...
+  };
+  // TODO: refactor so it is Typescript valid
+  const filterShippingSettings = (s: ShopShippingInterface) => (
+    (s.collection !== ShopShippingCollectionType.CARRIERS || (!!s.properties.active && !s.properties.deleted))
+      && (!s.properties.country_ids || settings.targetCountries?.some((tc: string) => s.properties.country_ids.includes(tc))));
+
+  switch (settings.shippingSetup) {
+    case ShippingSetupOption.ESTIMATE:
+      productFeedApiPayload.rate = settings.rate;
+      productFeedApiPayload.estimateCarriers = settings.estimateCarriers?.map(filterEstimateCarriers);
```
Comment (Quetzacoalt91) [inline _dev/apps/ui/src/store/modules/product-feed/actions.ts:49]: If we want to improve the types used in the product feed configuration, we could make this attribute mandatory when the Estimate is chosen (no undefined

File: _dev/apps/ui/src/store/modules/product-feed/actions.ts [hunk 7]
```diff
@@ -1,54 +1,81 @@
...
-          || settings.targetCountries.some((tc: string) => s.properties.country_ids.includes(tc)))),
-      ),
-      additionalShippingSettings: settings.additionalShippingSettings,
-    } : {}
-  ),
-  attributeMapping: formatMappingToApi(settings.attributeMapping),
-  selectedProductCategories: settings.selectedProductCategories,
-  requestSynchronizationNow: settings.requestSynchronizationNow,
-});
+      productFeedApiPayload.shippingSettings = settings.shippingSettings?.filter(filterShippingSettings);
```
Comment (Quetzacoalt91) [inline _dev/apps/ui/src/store/modules/product-feed/actions.ts:53]: If we want to improve the types used in the product feed configuration, we could make this attribute mandatory when the Import is chosen.

File: _dev/apps/ui/src/store/modules/product-feed/state.ts [hunk 8]
```diff
@@ -34,15 +34,25 @@ export type ProductFeedReport = {
 }
 export interface ProductFeedSettings {
   shippingSetup: ShippingSetupOption|null;
-  estimateCarriers: CustomCarrier[];
-  shippingSettings: ShopShippingInterface[];
-  rate: RateType|null;
-  deliveryDetails: DeliveryDetail[];
+  estimateCarriers?: CustomCarrier[];
+  shippingSettings?: ShopShippingInterface[];
+  rate?: RateType|null;
```
Comment (Quetzacoalt91) [inline _dev/apps/ui/src/store/modules/product-feed/state.ts:39]: What's the reason behind this change?

File: _dev/apps/ui/src/store/modules/product-feed/state.ts [hunk 9]
```diff
@@ -34,15 +34,25 @@ export type ProductFeedReport = {
 }
 export interface ProductFeedSettings {
   shippingSetup: ShippingSetupOption|null;
-  estimateCarriers: CustomCarrier[];
-  shippingSettings: ShopShippingInterface[];
-  rate: RateType|null;
-  deliveryDetails: DeliveryDetail[];
+  estimateCarriers?: CustomCarrier[];
+  shippingSettings?: ShopShippingInterface[];
```
Comment (Quetzacoalt91) [inline _dev/apps/ui/src/store/modules/product-feed/state.ts:38]: These variable have been made mandatory in the vue store to prevent issues with the reactiveness of the UI (going from undefined to a value prevents us to see the form being updated).

If you made this change for the API payload, don't you think it should go to its own type?

File: _dev/apps/ui/src/store/modules/product-feed/actions.ts (hunk 6 above)
Comment (kktos) [reply _dev/apps/ui/src/store/modules/product-feed/actions.ts:48]: ok. but it seems to be set only if `settings.shippingSetup === ShippingSetupOption.ESTIMATE`
If I remove the optional flag on this field, that means, I'll have to always set it to a value, [].
Here :
```typescript
export const createProductFeedApiPayload = (settings: ProductFeedAPIPayload): ProductFeedAPIPayload => {
  const productFeedApiPayload: ProductFeedAPIPayload = {
    autoImportTaxSettings: settings.autoImportTaxSettings,
    shippingSetup: settings.shippingSetup,
    targetCountries: settings.targetCountries,

    // always need to set it
    estimateCarriers: [],

    attributeMapping: formatMappingToApi(settings.attributeMapping as any),
    selectedProductCategories: settings.selectedProductCategories,
    requestSynchronizationNow: settings.requestSynchronizationNow,
  };
```
if that's ok with you, I'll make the change.

File: _dev/apps/ui/src/store/modules/product-feed/actions.ts (hunk 7 above)
Comment (kktos) [reply _dev/apps/ui/src/store/modules/product-feed/actions.ts:52]: same as for `estimateCarriers`.
The `shippingSettings` is set only if `settings.shippingSetup === ShippingSetupOption.IMPORT`
As for the previous one, if that's ok, I'll make the change

File: _dev/apps/ui/src/store/modules/product-feed/state.ts (hunk 9 above)
Comment (kktos) [reply _dev/apps/ui/src/store/modules/product-feed/state.ts:38]: ok. so you're answering the question I asked in another comment.
As these fields must always be present, that means I need to add a default value.

File: _dev/apps/ui/src/store/modules/product-feed/state.ts (hunk 8 above)
Comment (kktos) [reply _dev/apps/ui/src/store/modules/product-feed/state.ts:39]: that was because of how things were done in `createProductFeedApiPayload`

File: _dev/apps/ui/src/store/modules/product-feed/actions.ts (hunk 6 above)
Comment (Quetzacoalt91) [reply _dev/apps/ui/src/store/modules/product-feed/actions.ts:48]: I think we should have two completely different types to be safe. One for the state in which some attributes are always define, one for the API payload in which you put some keys as optional.
